framework/ui/view/qmllaunchprovider.cpp

```diff
@@ -137,6 +141,61 @@ RetVal<Val> QmlLaunchProvider::toRetVal(const QVariant& jsrv) const
     return rv;
 }
 
+RetVal<QmlLaunchProvider::OpenData> QmlLaunchProvider::openWidget(const UriQuery &q)
+{
+    RetVal<OpenData> result;
+
+    QString uri = QString::fromStdString(q.uri().toString());
+
+    int widgetMetaTypeId = uriRegister()->widgetDialogMetaTypeId(uri);
+    QString objectId = QString::number(widgetMetaTypeId); // unque?
+
+    void *widgetClassPtr = QMetaType::create(widgetMetaTypeId);
+    QDialog* dialog = static_cast<QDialog*>(widgetClassPtr);
+
+    if (!dialog) {
+        result.ret = make_ret(Ret::Code::UnknownError);
+        return result;
+    }
+
+    QVariantMap status;
+    connect(dialog, &QDialog::finished, [=, &status, &objectId](int finishStatus){
+        status["errcode"] = static_cast<int>(Ret::Code::Ok);
+        status["value"] = finishStatus;
+
+        onPopupClose(objectId, status);
+    });
+
+    onOpen(PAGE_TYPE_WIDGET);
+
+    dialog->exec();
+
+    QMetaType::destroy(widgetMetaTypeId, widgetClassPtr);
+
+    result.ret = toRet(status);
+    result.val.sync = false; // TODO
+    result.val.objectID = QString::number(widgetMetaTypeId);
+
+    return result;
+}
+
+RetVal<QmlLaunchProvider::OpenData> QmlLaunchProvider::openQml(const UriQuery &q)
```

Comment:
**Qml**LaunchProvider::openQML
**Qml**LaunchProvider::openWidget

Isn't it better to call it just a "LaunchProvider"?